src/Symfony/Component/Messenger/Middleware/StackMiddleware.php

```diff
@@ -20,31 +20,74 @@
  */
 class StackMiddleware implements MiddlewareInterface, StackInterface
 {
-    private $middlewareIterator;
+    private $stack;
+    private $offset = 0;
 
-    public function __construct(\Iterator $middlewareIterator = null)
+    /**
+     * @param iterable|MiddlewareInterface[]|MiddlewareInterface|null $middlewareIterator
+     */
+    public function __construct($middlewareIterator = null)
     {
-        $this->middlewareIterator = $middlewareIterator;
+        $this->stack = new MiddlewareStack();
+
+        if (null === $middlewareIterator) {
+            return;
+        }
+
+        if ($middlewareIterator instanceof \Iterator) {
+            $this->stack->iterator = $middlewareIterator;
+        } elseif ($middlewareIterator instanceof MiddlewareInterface) {
+            $this->stack->stack[] = $middlewareIterator;
+        } elseif (!\is_iterable($middlewareIterator)) {
+            throw new \TypeError(sprintf('Argument 1 passed to %s() must be iterable of %s, %s given.', __METHOD__, MiddlewareInterface::class, \is_object($middlewareIterator) ? \get_class($middlewareIterator) : \gettype($middlewareIterator)));
+        } else {
+            $this->stack->iterator = (function () use ($middlewareIterator) {
+                yield from $middlewareIterator;
+            })();
+        }
     }
 
     public function next(): MiddlewareInterface
     {
-        if (null === $iterator = $this->middlewareIterator) {
+        if (null === $next = $this->stack->next($this->offset)) {
             return $this;
         }
-        $iterator->next();
 
-        if (!$iterator->valid()) {
-            $this->middlewareIterator = null;
+        ++$this->offset;
 
-            return $this;
-        }
-
-        return $iterator->current();
+        return $next;
     }
 
     public function handle(Envelope $envelope, StackInterface $stack): Envelope
     {
         return $envelope;
     }
 }
+
+/**
+ * @internal
+ */
+class MiddlewareStack
+{
+    public $iterator;
+    public $stack = [];
+
+    public function next(int $offset): ?MiddlewareInterface
+    {
+        if (isset($this->stack[$offset])) {
+            return $this->stack[$offset];
+        }
+
+        if (null === $this->iterator) {
+            return null;
+        }
+
+        $this->iterator->next();
+
+        if (!$this->iterator->valid()) {
+            return $this->iterator = null;
+        }
+
+        return $this->stack[] = $this->iterator->current();
```

Comment:
I had to stare for a minute at this. When you clone the `StackMiddleware`, each clone will still maintain this same, one instance of this `MiddlewareStack`, but will now maintain their own `StackMiddleware.offset` property (so, that property will move independently). The first `StackMiddleware` that uses an item from the iterator will use it, but store it on the stack so that it's available when the other `StackMiddleware` asks for that same `$offset`.